### DISCUSSION TITLE
xtask/CI: Clarify the vision

### DISCUSSION THREAD

**Discussion by o0Ignition0o** [apollographql/router#160] [Implementation]
Looking at various comments about the [xtask steps](https://github.com/apollographql/router/pull/89#discussion_r751029442) and [circleci build times efforts](https://github.com/apollographql/router/pull/147#issuecomment-971837124) we might not share a common vision for:

- What xtask is supposed to do
- The CI requirements
- The scope of both

From this gathering of requirements and goals should emerge a set of rules and a clearer way for us to decide on the tradeoffs.

For example the CI build can run many scenarios in parallel, without being stuck `Blocking waiting for file lock on build directory`, thus decreasing the feedback loop. 
Xtask provides a more fine grained control over the environment, and the available capabilities at hand. Moreover xtask commands can be run locally _and_ in CI, which is a great plus.

**Comment by cecton** [comment]
`xtask` is helpful for:
- avoiding common pitfalls in bash scripts
- moving away logic from CI specifics

I would prefer to keep most (if not everything we can) under xtask so it is easy to run things locally. In the circleci yaml file we should only have things proper to circle ci itself.

In this case you want to parallelize. To do that you will need to:
1. build most of the dependencies once
2. cache it
3. then have multiple CI task that use that cache and run every feature combination.

Unfortunately with the declarative nature of the yaml, you need to specify the feature combinations on both side. One way to circumvent this issue would be to create an xtask command `generate-circle-ci-yaml` that will generate the file automatically with all the feature sets.
